packages/deployer/src/compiler.ts

```diff
@@ -116,67 +150,51 @@ export class Compiler {
     public async compileAllAsync(): Promise<void> {
         await this._createArtifactsDirIfDoesNotExistAsync();
         this._contractSourcesIfExists = await Compiler._getContractSourcesAsync(this._contractsDir);
-        const contractBaseNames = _.keys(this._contractSourcesIfExists);
-        const compiledContractPromises = _.map(contractBaseNames, async (contractBaseName: string): Promise<void> => {
-            return this._compileContractAsync(contractBaseName);
+        _.forIn(this._contractSourcesIfExists, (source, fileName) => {
+            this._contractSourceData[fileName] = Compiler._getContractSpecificSourceData(source);
         });
-        await Promise.all(compiledContractPromises);
-
+        const fileNames = this._specifiedContracts.has(ALL_CONTRACTS_IDENTIFIER)
+            ? _.keys(this._contractSourcesIfExists)
+            : Array.from(this._specifiedContracts.values());
+        _.forEach(fileNames, fileName => {
+            this._setSourceTreeHash(fileName);
+        });
+        await Promise.all(_.map(fileNames, async fileName => this._compileContractAsync(fileName)));
         this._solcErrors.forEach(errMsg => {
             utils.consoleLog(errMsg);
         });
     }
     /**
      * Compiles contract and saves artifact to artifactsDir.
-     * @param contractBaseName Name of contract with '.sol' extension.
+     * @param fileName Name of contract with '.sol' extension.
      */
-    private async _compileContractAsync(contractBaseName: string): Promise<void> {
+    private async _compileContractAsync(fileName: string): Promise<void> {
         if (_.isUndefined(this._contractSourcesIfExists)) {
             throw new Error('Contract sources not yet initialized');
         }
+        const contractSpecificSourceData = this._contractSourceData[fileName];
+        const currentArtifact = (await this._getContractArtifactOrReturnAsync(fileName)) as ContractArtifact;
```

Comment:
Rename: `_getContractArtifactOrReturnAsync` to `_getContractArtifactIfExistsAsync` and `currentArtifact` to `currentArtifactIfExists`